src/test/java/com/puppycrawl/tools/checkstyle/utils/ASTUtils.java

```diff
@@ -0,0 +1,81 @@
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.utils;
+
+import java.io.File;
+import java.io.IOException;
+
+import antlr.ANTLRException;
+import com.puppycrawl.tools.checkstyle.TreeWalker;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FileContents;
+import com.puppycrawl.tools.checkstyle.api.FileText;
+
+public final class ASTUtils {
+    private ASTUtils() {
+    }
+
+    /**
+     * Creates a new <code>File</code> instance from the <code>filename</code> relative to
+     * src/test/resources/com/puppycrawl/tools/checkstyle/utils <code>folder</code>.
+     *
+     * @param filename filename
+     * @return The abstract pathname denoting file with <code>filename</code> situated in
+     * the <code>folder</code>.
+     */
+    public static File getFile(String filename) {
+        return new File("src/test/resources/com/puppycrawl/tools/checkstyle/utils/" + filename);
+    }
+
+    /**
+     * Parses Java source file. Result AST contains comment nodes.
+     *
+     * @param file file to parse
+     * @return DetailAST tree
+     * @throws NullPointerException if the text is null
+     * @throws IOException          if the file could not be read
+     * @throws ANTLRException       if parser or lexer failed
+     */
+    public static DetailAST parseFile(File file) throws IOException, ANTLRException {
+        final FileText text = new FileText(file.getAbsoluteFile(), "UTF-8");
+        final FileContents contents = new FileContents(text);
+        return TreeWalker.parseWithComments(contents);
+    }
+
+    /**
+     * Finds first node of <code>type</code> in <code>root</code> specified.
+     *
+     * @param root root tree node to begin search with
+     * @param type type of the node to search
+     * @return DetailAST tree with the node of <code>type</code> as a root
+     * or {@code null} if the <code>root</code> tree cantains no node of
+     * <code>type</code> specified.
+     */
+    public static DetailAST findFirstNode(DetailAST root, int type) {
```

Comment:
Ca we use [findFirstTokenByPredicate](https://github.com/MEZk/checkstyle/blob/257d13b6bb53245d2491e50d4ca8e39ddfc80035/src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtils.java#L200-L200) instead to avoid recuresion?